Microservice Product Search

Miroslav-Stopka opened the pull request:
**DO NOT MERGE**, this PR is for code review only

Experimental microservice for product search, integrated with the Shopsys Framework.
It uses Postgres at this point, but it should feature Elasticsearch in the near future.

| Q             | A
| ------------- | ---
|Description, reason for the PR| microservice approach to building large apps proves promising
|New feature| Yes <!-- Do not forget to update CHANGELOG.md and possibly docs/ -->
|BC breaks| Yes <!-- Do not forget to update UPGRADE.md -->
|Fixes issues| ...
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

PetrHeinz commented on microservices/product-search/config/packages/framework.yaml:
```diff
@@ -0,0 +1,30 @@
+framework:
+    secret: '%env(APP_SECRET)%'
+    #default_locale: en
```
I would suggest keeping them until someone changes the configuration to suit our needs. We haven't reconfigured it in any way, I think, so we kept the defaults even with the comments.

PetrHeinz replied:
Maybe we should actually configure it and remove the unnecessary comment right away.

Miroslav-Stopka replied:
Done :)

PetrHeinz commented on docker/conf/docker-compose-mac.yml.dist:
```diff
@@ -41,6 +42,21 @@ services:
+        container_name: shopsys-framework-microservice-product-search
+        volumes:
+            - shopsys-framework-microservice-product-search-sync:/var/www/html
+            - ./project-base/docker/php-fpm/php-ini-overrides.ini:/usr/local/etc/php/php.ini
```
You're right.

Miroslav-Stopka replied:
Fixed :)

PetrHeinz commented on microservices/product-search/.gitignore:
```diff
@@ -0,0 +1,13 @@
+/composer.lock
+/symfony.lock
+
+###> symfony/framework-bundle ###
```
Yes, the first commit (e715d1dd26dd3ae46b7f23fd61854dbf9a6e9e77) was supposed to be a clean install, but it's true we did a few alterations on top of it (eg. adding composer.lock to .gitignore or setting up docker-compose templates).

PetrHeinz commented on microservices/product-search/src/Component/String/DatabaseSearching.php:
```diff
@@ -0,0 +1,31 @@
+<?php
```
We will remove those classes before the branch is merged to master as we will use Elastic instead.

Miroslav-Stopka replied:
Plus I don't see a reason for adding commentary about copying? Microservice has its own repository, it is managed separately without any direct link to changes in the Shopsys Framework.

vitek-rostislav replied:
@Miroslav-Stopka well I think it is a good practice when you copy-paste something into your code to mention the original source :slightly_smiling_face:

PetrHeinz commented on packages/framework/src/Model/Product/Search/ProductSearchRepository.php:
```diff
@@ -69,7 +69,11 @@ protected function getFoundProductIds(QueryBuilder $productQueryBuilder, $search
 
         if (!isset($this->foundProductIdsCache[$domainId][$searchText])) {
-            $foundProductIds = $this->client->search($domainId, $searchText)->productIds;
+            $searchResult = $this->microserviceProductSearchClient->get('product-ids', [
```
Well... using a specific method is the same as using a specific endpoint, right? I mean, we could implement `__call` method on the `MicroserviceClient` so `getProductIds()` is resolved to `get('product-ids')` so it's effectively the same...

I feel that using the microservice interface is on the same level as using SQL (database interface). Which should be in the repository.

vitek-rostislav replied:
> using a specific method is the same as using a specific endpoint

Well, yes and no :smile: If the endpoint changed, you would have to change the implementation in `ProductSearchRepository`. But if you used some method from classs responsible for just that, you would need to change it's implementation but the `ProductSearchRepository` would remain untouched... 

> I feel that using the microservice interface is on the same level as using SQL (database interface). Which should be in the repository.

I can agree with that :+1: 

You convinced me, this solution is #good_enough :slightly_smiling_face:

PetrHeinz commented on docs/installation/installation-using-docker-application-setup.md:
```diff
@@ -3,25 +3,48 @@
 
-## 1. Setup the microservice product search
+## 1. Setup the microservice for product search
+If you are not running the application inside the [Monorepo](/docs/introduction/monorepo.md), you'll have to clone the [repository of the microservice](https://github.com/shopsys/microservice-product-search) inside a separate directory and configure the container in your `docker-compose.yml`:
```
Well, we wanted to go the easy way as how we're gonna install and run the microservice in project repositories is yet to be determined. But you're right it's inconsistent. What would you prefer? I would leave it as-is and refine the docs when we're clear how to install and run it easily.

vitek-rostislav replied:
I would just omit the beginning of the sentence - "If you are not running the application inside the monorepo" (as this is given somehow) and simply start with "You have to clone...". and maybe add Note below stating that this step is not necessary when running the app from monorepo

Miroslav-Stopka replied:
Slightly changed in fixup :)

PetrHeinz commented on docs/installation/installation-using-docker-application-setup.md:
```diff
@@ -3,25 +3,48 @@
+*Note: You can switch the running process to run in the background by pressing `CTRL+Z` and executing `bg`.*
+
 ## 2. Setup the Shopsys Framework application
 Now that the Docker environment is prepared we can setup the application itself.
```
Good idea.

Miroslav-Stopka replied:
Done in 51ec5cb4d65a837c45cf9c2c88ca97362e6de186

vitek-rostislav replied:
Well not exactly, I have added comment in the fixup commit

PetrHeinz commented on docs/installation/native-installation.md:
```diff
@@ -50,6 +51,25 @@ monolog:
+For the product search to be working correctly, you'll have to install a [microservice for product search](https://github.com/shopsys/microservice-product-search).
+It will act as a fully independent unit with a separate web server and repository.
+
+Just clone the repository into a separate directory, install all its dependencies, configure connection to the Postgres by and run the server.
```
Rather "...connection to the Postgres by *setting up the PG_CONNECTION_STRING environment variable (or the [.env file](http://symfony.com/doc/current/components/dotenv.html))* and run..."

Miroslav-Stopka replied:
Done in 51ec5cb4d65a837c45cf9c2c88ca97362e6de186

PetrHeinz commented on microservices/product-search/README.md:
```diff
@@ -2,27 +2,32 @@
-For a call the microservice from the Shopsys Framework based application, there is already prepared a component [MicroserviceClient](packages/framework/src/Component/Microservice/MicroserviceClient.php).
-The instance of this component is parametrized with the url on which the microservice is available. 
+### MicroserviceClient component
+To call the microservice from the Shopsys Framework based application, there is already prepared a component [MicroserviceClient](packages/framework/src/Component/Microservice/MicroserviceClient.php).
```
"...there is *an* already prepared ~a~ component..."

Miroslav-Stopka replied:
Done in 51ec5cb4d65a837c45cf9c2c88ca97362e6de186

PetrHeinz commented on microservices/product-search/README.md:
```diff
@@ -0,0 +1,83 @@
+## Support
+What to do when you are in troubles or need some help? Best way is to contact us on our Slack [http://slack.shopsys-framework.com/](http://slack.shopsys-framework.com/)
+
+If you want to [report issues](https://github.com/shopsys/shopsys/issues/new), please use the main [Shopsys repository](https://github.com/shopsys/shopsys).
```
It's from our README.md boilerplate, we would have to change it everywhere.

Also, these are two separate use-cases: "I want to help!" / "I want help!", so it's not an issue if there is a duplicate info imo.

PetrHeinz commented:
Thanks @vitek-rostislav for the review and @scundall for the tips how to improve the microservice in the future.

To address the microservice being only requested by trusted clients - this is our current assumption and yes, it's ensured by having it inside the Docker network without any exposed ports. How to secure the connection to microservices in general is yet to be determined, in the meantime I'd suppose it's only accessible in a private network.

Miroslav-Stopka commented on microservices/product-search/src/Controller/SearchController.php:
```diff
@@ -0,0 +1,30 @@
+
+    public function productIdsAction(Request $request, int $domainId): JsonResponse
+    {
+        $query = $request->query->get('query', '');
```
You are right, query will be replaced by searchText :)

Miroslav-Stopka commented:
Hi @vitek-rostislav, thank you for your review :). I am sure it was not an easy one.

First, your notes:

- there is en empty .gitignore in src/Controller folder, is this necessary?
    - this file was necessary when that folder was empty. I removed it now in fixup 

- I am little bit constrained about the default Symfony comments in configuration files - I am not sure whether it is a good idea to keep them there after the app is configured for our needs, they may not be relevant anymore
     - I think some of them can be quite useful. I don't see a reason for removing valid commentary. But you are partially right, so I removed some useless entries from microservices/product-search/config/packages/framework.yaml

- one of the commit messages (I believe it is the logging one) states that the microservice is only requested by trusted clients. How is this ensured? By running it in docker container? I am little bit confused by the statement
     - @PetrHeinz commented this later :)

- I would appreciate if we used monospace for class names in docs (e.g. MicroserviceClient in README.md) - it would be more readable IMO
     - Done :)

Miroslav-Stopka commented on docs/installation/native-installation.md:
```diff
@@ -50,6 +51,25 @@ monolog:
+
+In this moment the microservice is ready for the requests processing.
+
+*Note: If you use other port for the microservice to run, you'll have to pass its URL to the application as a parameter.*
```
Done in 51ec5cb4d65a837c45cf9c2c88ca97362e6de186

Miroslav-Stopka commented:
@vitek-rostislav please perform tests, thank you.

vitek-rostislav commented:
Hi @Miroslav-Stopka, I have tested it nad it works! :tada: great job :+1: Last 2 notes
 - you should not remove the `.gitignore` in fixup commit of the first one but in the commit where you create the content of the folder (i.e. https://github.com/shopsys/shopsys/pull/368/commits/8b1b08573047fffe7c3c2b20fad3eb52931c8755)
- as I commented on the file, one sentence should be on another line :slightly_smiling_face: 

That is all, review and tests are done, please fix the last 2 things and you can process to the business validation :+1:

Miroslav-Stopka commented:
@vitek-rostislav I changed PR according your last two notices :)

Miroslav-Stopka commented:
@LukasHeinz done

Miroslav-Stopka commented:
Merged in duplicate branch #401
